web3/contract.py

```diff
@@ -1328,6 +1332,39 @@ def factory(cls, class_name, **kwargs):
         return PropertyCheckingFactory(class_name, (cls,), kwargs)
 
 
+class CallerMethod:
+    def __init__(self, function, normalizers=None):
+        self._function = function
+
+    def __call__(self, *args, **kwargs):
+        # TODO: Remove this and consolidate into ContractCaller.
+        # I don't know why this is working but it is.
+        return self._function(*args, **kwargs).call()
+
+
+class ContractCaller:
+    def __init__(self, abi, web3, address, *args, **kwargs):
+        self.web3 = web3
+        self.address = address
+        self.abi = abi
+
+        if self.abi:
+            self._functions = filter_by_type('function', self.abi)
+            for func in self._functions:
+                fn = ContractFunction.factory(
+                            func['name'],
+                            web3=self.web3,
+                            contract_abi=self.abi,
+                            address=self.address,
+                            function_identifier=func['name'])
+
+                caller_method = CallerMethod(fn)
+                setattr(self, func['name'], caller_method)
+
+    def __call__(self, *args, **kwargs):
+        return type(self)(self.abi, self.web3, self.address, *args, **kwargs)
```

Comment:
This is where the kwargs could be consolidated into `transaction_dict`. I'm not sure what `args` would be here, besides the transaction dict.

Something like this could work:
```py
    def __call__(self, transaction_dict=None, **kwargs):
        assert transaction_dict is None or len(kwargs) == 0  # but not actually an assert
        return type(self)(self.abi, self.web3, self.address, transaction_dict=(transaction_dict or kwargs))

Reply:
IIUC the `**kwargs` portion is to support keyword based transaction parameters?  If so my previous comment stands about holding off on introducing this option into the API.

Reply:
It looks like you can use keyword args for the transaction parameters and `block_identifier`

Reply:
I'm not sure I'm keen on having the `block_identifier` be something that you set once in the caller and keep around. At the very least, it makes me less excited about the earlier idea (break up the transaction dict into keywords), because then you have a mess of transaction fields and other stuff like the `block_identifier`.

But... if we don't have the `block_identifier` in the caller, then is it just impossible to set the block identifier using the `ContractCaller`? (That might be okay, maybe the "long form" call is required to do that kind of request)